### PR TITLE
Data downloader

### PR DESCRIPTION
**Pull request by zphang**
Added a data downloader at `jiant/scripts/download_data/runscript.py`.

Currently supports:
  * All GLUE tasks (including diagnostic, SNLI)
  * All SuperGLUE tasks (including AX-g, AX-b; MultiRC and ReCoRD are in `.jsonl` format which is not currently compatible with task implementations)
  * XTREME tasks (excluding UDPos, PAN-X, but including Bucc2018 and Tatoeba, that don't have task implementations yet)
  * Other: SNLI (nlp), SQuAD v1.1 (downloaded), SQuAD v2 (downloaded)

The goal will be to deprecate `export_glue_data.py` and potentially `write_task_configs.py` as well.

### CODE REVIEW

**Comment by pyeres** [comment]
@zphang and @jeswan, if you haven't already discussed the connection between this work and the work @jeswan is doing w/ Hugging Face's NLP, please connect (it looks like GLUE, SuperGLUE and XTREME data may be available there).

**Comment by zphang** [comment]
I think this will be useful in the short term independent of adding NLP support. As long as we are still relying on data files for some tasks, we should make it easy to download the relevant datasets (as well as a reference for what task configs look like).

File: jiant/utils/python/io.py [hunk 1]
```diff
@@ -79,3 +79,14 @@ def get_num_lines(path):
         for i, l in enumerate(f):
             pass
     return i + 1
+
+
+def get_dir(*args):
```
Comment (jeswan) [inline jiant/utils/python/io.py:84]: unclear function name

File: jiant/utils/python/datastructures.py [hunk 2]
```diff
@@ -105,6 +105,12 @@ def sort_dict(d: dict):
     return {k: d[k] for k in sorted(list(d.keys()))}
 
 
+def set_dict_keys(d, key_list):
+    """Return a new dictionary with specified key order"""
+    assert set(d) == set(key_list)
+    return {k: d[k] for k in key_list}
```
Comment (jeswan) [inline jiant/utils/python/datastructures.py:111]: insertion order is only guaranteed in Python 3.6

**Comment by jeswan** [comment]
Spoke offline to @zphang. He is working on implementing the nlp data downloader in this PR.

**Comment by zphang** [comment]
Updated the PR to switch GLUE and SuperGLUE tasks to NLP, also added SQuAD v1 and v2. Also as mentioned above, MultiRC and ReCoRD's loaders are not currently compatible with the task data loaders.

I'm postponing moving XTREME over to using NLP because it has a more complex pipeline, and also because the NLP version is kind of messy currently (e.g. inconsistent handling of language subsets):

* Please pick one among the available configs: ['XNLI', 'tydiqa', 'SQuAD', 'PAN-X.af', 'PAN-X.ar', 'PAN-X.bg', 'PAN-X.bn', 'PAN-X.de', 'PAN-X.el', 'PAN-X.en', 'PAN-X.es', 'PAN-X.et', 'PAN-X.eu', 'PAN-X.fa', 'PAN-X.fi', 'PAN-X.fr', 'PAN-X.he', 'PAN-X.hi', 'PAN-X.hu', 'PAN-X.id', 'PAN-X.it', 'PAN-X.ja', 'PAN-X.jv', 'PAN-X.ka', 'PAN-X.kk', 'PAN-X.ko', 'PAN-X.ml', 'PAN-X.mr', 'PAN-X.ms', 'PAN-X.my', 'PAN-X.nl', 'PAN-X.pt', 'PAN-X.ru', 'PAN-X.sw', 'PAN-X.ta', 'PAN-X.te', 'PAN-X.th', 'PAN-X.tl', 'PAN-X.tr', 'PAN-X.ur', 'PAN-X.vi', 'PAN-X.yo', 'PAN-X.zh', 'MLQA.ar.ar', 'MLQA.ar.de', 'MLQA.ar.vi', 'MLQA.ar.zh', 'MLQA.ar.en', 'MLQA.ar.es', 'MLQA.ar.hi', 'MLQA.de.ar', 'MLQA.de.de', 'MLQA.de.vi', 'MLQA.de.zh', 'MLQA.de.en', 'MLQA.de.es', 'MLQA.de.hi', 'MLQA.vi.ar', 'MLQA.vi.de', 'MLQA.vi.vi', 'MLQA.vi.zh', 'MLQA.vi.en', 'MLQA.vi.es', 'MLQA.vi.hi', 'MLQA.zh.ar', 'MLQA.zh.de', 'MLQA.zh.vi', 'MLQA.zh.zh', 'MLQA.zh.en', 'MLQA.zh.es', 'MLQA.zh.hi', 'MLQA.en.ar', 'MLQA.en.de', 'MLQA.en.vi', 'MLQA.en.zh', 'MLQA.en.en', 'MLQA.en.es', 'MLQA.en.hi', 'MLQA.es.ar', 'MLQA.es.de', 'MLQA.es.vi', 'MLQA.es.zh', 'MLQA.es.en', 'MLQA.es.es', 'MLQA.es.hi', 'MLQA.hi.ar', 'MLQA.hi.de', 'MLQA.hi.vi', 'MLQA.hi.zh', 'MLQA.hi.en', 'MLQA.hi.es', 'MLQA.hi.hi', 'XQuAD.ar', 'XQuAD.de', 'XQuAD.vi', 'XQuAD.zh', 'XQuAD.en', 'XQuAD.es', 'XQuAD.hi', 'XQuAD.el', 'XQuAD.ru', 'XQuAD.th', 'XQuAD.tr', 'bucc18.de', 'bucc18.fr', 'bucc18.zh', 'bucc18.ru', 'PAWS-X.de', 'PAWS-X.en', 'PAWS-X.es', 'PAWS-X.fr', 'PAWS-X.ja', 'PAWS-X.ko', 'PAWS-X.zh', 'tatoeba.afr', 'tatoeba.ara', 'tatoeba.ben', 'tatoeba.bul', 'tatoeba.deu', 'tatoeba.cmn', 'tatoeba.ell', 'tatoeba.est', 'tatoeba.eus', 'tatoeba.fin', 'tatoeba.fra', 'tatoeba.heb', 'tatoeba.hin', 'tatoeba.hun', 'tatoeba.ind', 'tatoeba.ita', 'tatoeba.jav', 'tatoeba.jpn', 'tatoeba.kat', 'tatoeba.kaz', 'tatoeba.kor', 'tatoeba.mal', 'tatoeba.mar', 'tatoeba.nld', 'tatoeba.pes', 'tatoeba.por', 'tatoeba.rus', 'tatoeba.spa', 'tatoeba.swh', 'tatoeba.tam', 'tatoeba.tgl', 'tatoeba.tha', 'tatoeba.tur', 'tatoeba.urd', 'tatoeba.vie', 'udpos.Afrikaans', 'udpos.Arabic', 'udpos.Basque', 'udpos.Bulgarian', 'udpos.Dutch', 'udpos.English', 'udpos.Estonian', 'udpos.Finnish', 'udpos.French', 'udpos.German', 'udpos.Greek', 'udpos.Hebrew', 'udpos.Hindi', 'udpos.Hungarian', 'udpos.Indonesian', 'udpos.Italian', 'udpos.Japanese', 'udpos.Kazakh', 'udpos.Korean', 'udpos.Chinese', 'udpos.Marathi', 'udpos.Persian', 'udpos.Portuguese', 'udpos.Russian', 'udpos.Spanish', 'udpos.Tagalog', 'udpos.Tamil', 'udpos.Telugu', 'udpos.Thai', 'udpos.Turkish', 'udpos.Urdu', 'udpos.Vietnamese', 'udpos.Yoruba']

My recommendation is have the flexibility for both NLP and download-based data sets for now, using NLP where it's obvious and convenient, and downloading otherwise.

File: jiant/utils/python/io.py (hunk 1 above)
Comment (zphang) [reply jiant/utils/python/io.py]: Renamed to `create_dir`

File: jiant/utils/python/datastructures.py (hunk 2 above)
Comment (zphang) [reply jiant/utils/python/datastructures.py:111]: We're currently using dataclasses so we require at least 3.7.

**Comment by jeswan** [comment]
@zphang Seems important to have the ability to list the supported datasets in the downloader. Can you add this? Also, the `download_data_and_write_config` style of functions prevent writing configs in a centralized, abstract-able way. Would prefer if downloading data and config generation were decoupled.

**Comment by zphang** [comment]
I'll add something to list the available tasks. (I was considering using a registry, but I'm inclined to keep things simple for now.)

Regarding download/writing configs, I'm not sure how to do this effectively. The configs are tied to the file-system organization of data (as well as the specific paths), as well as the task-specific configurations, so it's hard to abstract away just the config-writing portion without needing to provide most of the same information. `write_task_configs.py` was an attempt to abstract it away, but I think jointly writing the data and task configs is a much intuitive workflow. (In theory the task configs should be more closely linked to the task definition, but that'll require some more reorganization work.)

Open to ideas though.

**Comment by jeswan** [comment]
@zphang Thanks for your thoughts. I'm planning to checkout the branch to attempt to play with/implement some ideas.

File: jiant/scripts/download_data/datasets/xtreme.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+    for lang in languages:
+        task_name = f"bucc2018_{lang}"
+        task_data_path = py_io.create_dir(task_data_base_path, task_name)
+        val_eng_path = os.path.join(task_data_path, f"{lang}-en.dev.en")
+        val_other_path = os.path.join(task_data_path, f"{lang}-en.dev.{lang}")
+        val_labels_path = os.path.join(task_data_path, f"{lang}-en.dev.gold")
+        test_eng_path = os.path.join(task_data_path, f"{lang}-en.test.en")
+        test_other_path = os.path.join(task_data_path, f"{lang}-en.test.{lang}")
+        # sample -> dev
+        # training -> test (yup, it's weird)
```
Comment (jeswan) [inline jiant/scripts/download_data/datasets/xtreme.py:218]: what lines do these comments refer to?

**Comment by jeswan** [comment]
@zphang Made some suggested modifications here: https://github.com/jiant-dev/jiant/commit/2845baf7bc46798a028555404b5f3c3ee0b137eb. Let me know what you think and we can cherry-pick the changes.

File: jiant/scripts/download_data/datasets/xtreme.py (hunk 3 above)
Comment (zphang) [reply jiant/scripts/download_data/datasets/xtreme.py:218]: Bucc2018 doesn't have actual dev/test sets. Instead, the "sample" set is used as dev, and the "training" set is used as the test set. This comment explains the remapping.

**Comment by jeswan** [comment]
Is there anything preventing export_glue_data.py and write_task_configs.py from being removed?

File: jiant/scripts/download_data/datasets/nlp_tasks.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    "cola": {
+        "path": "glue",
+        "name": "cola",
+        "field_map": {"sentence": "text"},
+        "label_map": {0: "0", 1: "1"},
+    },
+    "mnli": {
+        "path": "glue",
+        "name": "mnli",
+        "label_map": {0: "contradiction", 1: "entailment", 2: "neutral"},
```
Comment (jeswan) [inline jiant/scripts/download_data/datasets/nlp_tasks.py:18]: Looks like `label_map` must match the task ID_TO_LABEL. Seems like `label_map` should be changed to `MnliTask.ID_TO_LABEL`.

File: jiant/scripts/download_data/datasets/nlp_tasks.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+import jiant.scripts.download_data.utils as download_utils
+import jiant.utils.python.io as py_io
+
+
+NLP_CONVERSION_DICT = {
+    # === GLUE === #
+    "cola": {
+        "path": "glue",
+        "name": "cola",
+        "field_map": {"sentence": "text"},
```
Comment (jeswan) [inline jiant/scripts/download_data/datasets/nlp_tasks.py:12]: did `nlp` change the name of the field from the original dataset?

File: jiant/scripts/download_data/datasets/nlp_tasks.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    },
+    "superglue_winogender_diagnostics": {
+        "path": "super_glue",
+        "name": "axg",
+        "label_map": {0: "entailment", 1: "not_entailment"},
+        "phase_map": None,
+        "jiant_task_name": "rte",
+    },
+    # === Other === #
+    "snli": {"path": "snli", "label_map": {0: "contradiction", 1: "entailment", 2: "neutral"}},
```
Comment (jeswan) [inline jiant/scripts/download_data/datasets/nlp_tasks.py:105]: snli needs "name" and has an error

File: jiant/scripts/download_data/datasets/nlp_tasks.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        "field_map": {"sentence1": "premise", "sentence2": "hypothesis"},
+        "label_map": {0: "entailment", 1: "not_entailment"},
+        "phase_map": None,
+        "jiant_task_name": "rte",
+    },
+    "superglue_winogender_diagnostics": {
+        "path": "super_glue",
+        "name": "axg",
+        "label_map": {0: "entailment", 1: "not_entailment"},
+        "phase_map": None,
```
Comment (jeswan) [inline jiant/scripts/download_data/datasets/nlp_tasks.py:101]: can this be removed?

File: jiant/scripts/download_data/datasets/nlp_tasks.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    "multirc": {"path": "super_glue", "name": "multirc"},
+    "record": {"path": "super_glue", "name": "record"},
+    "wic": {"path": "super_glue", "name": "wic", "label_map": {0: False, 1: True}},
+    "wsc": {"path": "super_glue", "name": "wsc.fixed", "label_map": {0: False, 1: True}},
+    "superglue_broadcoverage_diagnostics": {
+        "path": "super_glue",
+        "name": "axb",
+        "field_map": {"sentence1": "premise", "sentence2": "hypothesis"},
+        "label_map": {0: "entailment", 1: "not_entailment"},
+        "phase_map": None,
```
Comment (jeswan) [inline jiant/scripts/download_data/datasets/nlp_tasks.py:94]: can this be removed?

File: jiant/scripts/download_data/datasets/xtreme.py [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+    xnli_temp_path = py_io.create_dir(task_data_base_path, "xnli_temp")
+    download_utils.download_and_unzip(
+        "https://dl.fbaipublicfiles.com/XNLI/XNLI-1.0.zip", xnli_temp_path,
+    )
+    full_val_data = py_io.read_jsonl(os.path.join(xnli_temp_path, "XNLI-1.0", "xnli.dev.jsonl"))
+    val_data = datastructures.group_by(full_val_data, key_func=lambda elem: elem["language"])
+    full_test_data = py_io.read_jsonl(os.path.join(xnli_temp_path, "XNLI-1.0", "xnli.test.jsonl"))
+    test_data = datastructures.group_by(full_test_data, lambda elem: elem["language"])
+    languages = sorted(list(val_data))
+    assert len(languages) == 15
```
Comment (jeswan) [inline jiant/scripts/download_data/datasets/xtreme.py:20]: where does 15 come from?

File: jiant/scripts/download_data/datasets/xtreme.py [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+        other_out = os.path.join(task_data_path, f"{lang}-en.{lang}")
+        labels_out = os.path.join(task_data_path, f"{lang}-en.labels")
+        tgts = [line.strip() for line in py_io.read_file_lines(eng_src)]
+        os.rename(src=other_src, dst=other_out)
+        idx = range(len(tgts))
+        data = zip(tgts, idx)
+        with open(eng_out, "w") as ftgt, open(labels_out, "w") as flabels:
+            for t, i in sorted(data, key=lambda x: x[0]):
+                ftgt.write(f"{t}\n")
+                flabels.write(f"{i}\n")
```
Comment (jeswan) [inline jiant/scripts/download_data/datasets/xtreme.py:326]: I'm not sure what is going on here. Is this step obvious to someone familiar with the task?

File: jiant/scripts/download_data/datasets/xtreme.py [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+    languages = "de fr ru zh".split()
+    for lang in languages:
+        download_utils.download_and_untar(
+            f"https://comparable.limsi.fr/bucc2018/bucc2018-{lang}-en.training-gold.tar.bz2",
+            bucc2018_temp_path,
+        )
+        download_utils.download_and_untar(
+            f"https://comparable.limsi.fr/bucc2018/bucc2018-{lang}-en.sample-gold.tar.bz2",
+            bucc2018_temp_path,
+        )
```
Comment (jeswan) [inline jiant/scripts/download_data/datasets/xtreme.py:208]: looks like duplicate lines.

File: jiant/scripts/download_data/datasets/nlp_tasks.py (hunk 4 above)
Comment (zphang) [reply jiant/scripts/download_data/datasets/nlp_tasks.py:18]: Good point. (Should I be working off your commits now, or do you want to make these changes?)

File: jiant/scripts/download_data/datasets/nlp_tasks.py (hunk 5 above)
Comment (zphang) [reply jiant/scripts/download_data/datasets/nlp_tasks.py:12]: (For some of these, including CoLA) the datasets do not have column headers. So the naming is arbitrary on both our/HF's parts.)

For now, I'm mapping to the field names that the Task objects are expecting. We can consider switching over to using HF's naming conventions in a later PR, but this PR should not touch any of the task implementations.

File: jiant/scripts/download_data/datasets/nlp_tasks.py (hunk 6 above)
Comment (zphang) [reply jiant/scripts/download_data/datasets/nlp_tasks.py:105]: Good point. Will switch to `.get('name')`.

File: jiant/scripts/download_data/datasets/nlp_tasks.py (hunk 7 above)
Comment (zphang) [reply jiant/scripts/download_data/datasets/nlp_tasks.py:101]: Not really. (This parts a little annoying: we need to map "validation"->"valid", so the default is a `DEFAULT_PHASE_MAP`, but the diagnostic tasks do not have a validation set so we need to pass `phase_map=None`. The current solution is kind of ugly but at least it's sort of an explicit override.)

File: jiant/scripts/download_data/datasets/nlp_tasks.py (hunk 8 above)
Comment (zphang) [reply jiant/scripts/download_data/datasets/nlp_tasks.py:94]: See above.

File: jiant/scripts/download_data/datasets/xtreme.py (hunk 9 above)
Comment (zphang) [reply jiant/scripts/download_data/datasets/xtreme.py:20]: Ah, this was for testing. Can remove.

File: jiant/scripts/download_data/datasets/xtreme.py (hunk 10 above)
Comment (zphang) [reply jiant/scripts/download_data/datasets/xtreme.py:326]: Not really, I should add an explanation for this. See also, the original preprocessing code: https://github.com/google-research/xtreme/blob/975d5ffb86be9aab0f010054b56f4204d790aa18/utils_preprocess.py#L390-L400

Tatoeba is a retrieval dataset where you have a set of sentences in English and another set in another language, and you need to match them. Also, it's a dataset that doesn't have training data, so it's pretty much evaluation only. However, the dataset is distributed with the sentences in order, i.e. the retrieval pairing is the sentence order. The XTREME authors basically intentionally scramble the order by sorting one of the two sets alphabetically. I'm following their recipe, but also retaining the labels for internal scoring.

In theory, we can do away with the whole scrambling thing if we trust the users to not (accidentally) cheat, but I think this is a good compromise between having something that looks like the XTREME processed outputs while still owning the logic for preparing the data ourselves.

Comment (jeswan) [reply jiant/scripts/download_data/datasets/xtreme.py:326]: Thanks for the explanation. Do you mind adding a comment based on your explanation?

File: jiant/scripts/download_data/datasets/nlp_tasks.py (hunk 4 above)
Comment (jeswan) [reply jiant/scripts/download_data/datasets/nlp_tasks.py:18]: I'll try pushing to the PR branch (will likely need your approval).

File: jiant/scripts/download_data/datasets/xtreme.py (hunk 10 above)
Comment (jeswan) [reply jiant/scripts/download_data/datasets/xtreme.py:326]: Is it possible to use the xtreme/utils_preprocess.py directly?

**Comment by jeswan** [comment]
@zphang Since the branch is on your fork, it is easier to `cherry-pick` my changes: https://github.com/jiant-dev/jiant/commit/2845baf7bc46798a028555404b5f3c3ee0b137eb and c1c52d9f932823f395b034deb6f1a85de4152404. Let me know if you have any issues!

**Comment by zphang** [comment]
Merged in the changes, + ID_TO_LABEL. I made some minor tweaks to the CLI interface.

**Pull request merged**